### PR TITLE
Thunks: Make xcb's callback more robust.

### PR DESCRIPTION
**Pull request by Sonicadvance1**
Fixes a crash in xcb thunks since more stuff on FEX side has moved over to jemalloc.

Destructors don't actually get called when a shared library is removed. It's some weirdo quirk that we can't work around. Instead refcount Display connections being created and disconnected. Creating the thread on the first display creation, and tearing down on final display teardown.

Must be merged before #2564 otherwise that PR will break thunks.

### CODE REVIEW

File: ThunkLibs/libxcb/libxcb_Guest.cpp [hunk 1]
```diff
@@ -117,17 +129,33 @@ extern "C" {
 
   xcb_connection_t * xcb_connect(const char * a_0,int * a_1){
     auto ret = fexfn_pack_xcb_connect(a_0, a_1);
+
+    if (xcb_get_file_descriptor(ret) != -1) {
```
Comment (neobrain) [inline ThunkLibs/libxcb/libxcb_Guest.cpp:133]: XCB documentation suggests `xcb_connection_has_error` should be used for error checking.

File: ThunkLibs/libxcb/libxcb_Guest.cpp [hunk 2]
```diff
@@ -30,6 +30,8 @@ tags: thunklibs|xcb
 
 static std::thread CBThread{};
 static std::atomic<bool> CBDone{false};
+static std::mutex CBThreadMutex;
+static std::atomic<uint32_t> ScreenRefCount;
```
Comment (neobrain) [inline ThunkLibs/libxcb/libxcb_Guest.cpp:34]: This doesn't need to be atomic since it's guarded by a mutex.

**Review by neobrain — APPROVED** [review]
The general observation that destructors don't get called on library unload is concerning... can we hook into `dlclose` to trigger a manual cleanup for other libraries?

The XCB change itself looks fine.

File: ThunkLibs/libxcb/libxcb_Guest.cpp (hunk 1 above)
Comment (Sonicadvance1) [reply ThunkLibs/libxcb/libxcb_Guest.cpp:133]: We could use that on connect, but we can't use that on disconnect. I opted to use the same valid display connection testing for all three instances to be consistent. I could switch the two connect ones over if you want.

**Comment by Sonicadvance1** [comment]
> can we hook into dlclose to trigger a manual cleanup for other libraries?

Once we have redirected the FD for the thunk library to be loaded, we lose control of this. I haven't found a good workaround.

File: ThunkLibs/libxcb/libxcb_Guest.cpp (hunk 1 above)
Comment (neobrain) [reply ThunkLibs/libxcb/libxcb_Guest.cpp:133]: Fair enough. `xcb_get_file_descriptor` is fine then.

**Comment by neobrain** [comment]
Does compiling with `--fno-gnu-unique` cause destructors to be invoked on `dlclose` by any chance?

**Comment by neobrain** [comment]
Ah, bummer. Thanks for testing though!

**Pull request merged**